controlplane/kubeadm/controllers/upgrade_test.go

```diff
@@ -63,158 +64,52 @@ func TestKubeadmControlPlaneReconciler_upgradeControlPlane(t *testing.T) {
 	g.Expect(result).To(Equal(ctrl.Result{Requeue: true}))
 	g.Expect(err).NotTo(HaveOccurred())
 
-	machineList := &clusterv1.MachineList{}
-	g.Expect(fakeClient.List(context.Background(), machineList, client.InNamespace(cluster.Namespace))).To(Succeed())
-	g.Expect(machineList.Items).NotTo(BeEmpty())
-	g.Expect(machineList.Items).To(HaveLen(1))
+	// initial setup
+	initialMachine := &clusterv1.MachineList{}
+	g.Expect(fakeClient.List(context.Background(), initialMachine, client.InNamespace(cluster.Namespace))).To(Succeed())
+	g.Expect(initialMachine.Items).To(HaveLen(1))
 
-	machineCollection := internal.NewFilterableMachineCollection(&machineList.Items[0])
-	result, err = r.upgradeControlPlane(context.Background(), cluster, kcp, machineCollection, machineCollection, controlPlane)
+	// change the KCP spec so the machine becomes outdated
+	kcp.Spec.Version = "v1.17.4"
 
-	g.Expect(machineList.Items[0].Annotations).To(HaveKey(controlplanev1.SelectedForUpgradeAnnotation))
-
-	// TODO flesh out the rest of this test - this is currently least-effort to confirm a fix for an NPE when updating
-	// the etcd version
-	g.Expect(result).To(Equal(ctrl.Result{}))
+	// run upgrade the first time, expect we scale up
+	needingUpgrade := internal.NewFilterableMachineCollectionFromMachineList(initialMachine)
+	controlPlane.Machines = needingUpgrade
+	result, err = r.upgradeControlPlane(context.Background(), cluster, kcp, controlPlane)
+	g.Expect(result).To(Equal(ctrl.Result{Requeue: true}))
+	g.Expect(err).To(BeNil())
+	bothMachines := &clusterv1.MachineList{}
+	g.Expect(fakeClient.List(context.Background(), bothMachines, client.InNamespace(cluster.Namespace))).To(Succeed())
+	g.Expect(bothMachines.Items).To(HaveLen(2))
+
+	// run upgrade a second time, simulate that the node has not appeared yet but the machine exists
+	r.managementCluster.(*fakeManagementCluster).ControlPlaneHealthy = false
+	_, err = r.upgradeControlPlane(context.Background(), cluster, kcp, controlPlane)
 	g.Expect(err).To(Equal(&capierrors.RequeueAfterError{RequeueAfter: healthCheckFailedRequeueAfter}))
-}
-
-func TestSelectMachineForUpgrade(t *testing.T) {
```

Comment:
Can we add back this test case now that we have `selectMachineForScaleDown`?

Reply:
Added it back in scale_test.go